SIPS/sip-27.md

```diff
@@ -0,0 +1,70 @@
+---
+sip: 27
+title: Accounts Metadata Request
+status: Draft
+author: Olaf Tomalka (@ritave)
+created: 2024-09-17
+---
+
+## Abstract
+
+This SIP allows Snaps to retrieve metadata related to accounts that exist in the extension.
+
+## Motivation
+
+The intention of this SIP is to allow Snaps providing new accounts for new chains to be able to list all accounts when selecting one during transfers. Currently the Snap can only access their own accounts, and can't do more complex UI flows.
+
+The specific need for this SIP is to allow a Snap to create a send flow with an account picker, with an example design as follows.
+
+<img src="../assets/sip-27/snap-account-list.png" alt="Account picker flow" width="600"/>
+
+## Specification
+
+> Indented sections like this are considered non-normative.
+
+> Usage of `CAIP-N` specifications, where `N` is a number, are references to [Chain Agnostic Improvement Proposals](https://github.com/ChainAgnostic/CAIPs).
+
+### Language
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT",
+"SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and
+"OPTIONAL" written in uppercase in this document are to be interpreted as described in [RFC 2119](https://www.ietf.org/rfc/rfc2119.txt)
+
+### Snap Manifest
+
+This SIP specifies a permission named `keyring_listAccountsAll`. This permission grants a Snap the ability to retrieve account metadata through an RPC call.
+
+The permission is specified in `snap.manifest.json` as follows:
+
+```json
+{
+  "initialPermissions": {
+    "keyring_listAccountsAll": {}
+  }
+}
+```
+
+### RPC Method
+
+This SIP exposes a new RPC method called `keyring_listAccountsAll` with no additional parameters.
+
+The RPC call returns with the following data:
+
+```typescript
+type KeyringAccount = {
+  id: string; // An extension-specific unique ID
```

Comment:
Does it make sense to expose this? What do you envision we use it for?

Reply:
When you have two accounts with the same address (eg 2 different hardware wallets with the same seed), you need to differentiate them somehow.

Reply:
I feel like this just complicates things. The Snap has no information about how the address was created, right? Is it even relevant for the Snap?

Reply:
This provides a stable identity for an account.

For example, you could want to map accounts to a user editable notes about those accounts.

You can't use the name since it's user editable and can change, and you can't use address because there are multiple accounts with the same address.

By providing a stable identity (similar to React's key property) you can easily manipulate those accounts.